lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp

```diff
@@ -0,0 +1,164 @@
+//===-- OperatingSystemSwift.cpp -----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#if LLDB_ENABLE_SWIFT
+
+#include "OperatingSystemSwift.h"
+
+#include "Plugins/Process/Utility/ThreadMemory.h"
+#include "lldb/Core/Debugger.h"
+#include "lldb/Core/Module.h"
+#include "lldb/Core/PluginManager.h"
+#include "lldb/Target/Process.h"
+#include "lldb/Target/Thread.h"
+#include "lldb/Target/ThreadList.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/StructuredData.h"
+
+#include "swift/Threading/ThreadLocalStorage.h"
+
+#include <memory>
+
+using namespace lldb;
+using namespace lldb_private;
+
+LLDB_PLUGIN_DEFINE(OperatingSystemSwift)
+
+void OperatingSystemSwift::Initialize() {
+  PluginManager::RegisterPlugin(GetPluginNameStatic(),
+                                GetPluginDescriptionStatic(), CreateInstance,
+                                nullptr);
+}
+
+void OperatingSystemSwift::Terminate() {
+  PluginManager::UnregisterPlugin(CreateInstance);
+}
+
+OperatingSystem *OperatingSystemSwift::CreateInstance(Process *process,
+                                                      bool force) {
+  if (!process)
+    return nullptr;
+
+  /// Heuristic to only load the plugin for swift programs.
+  FileSpec swift_module_name = FileSpec("libswiftCore.dylib");
```

Comment:
This only works in a very narrow subset of Swift programs. You should defer to SwiftLanguageRuntime's helper code.
The cases this doesn't support are: non-Darwin (".dylib"), statically linked Swift stdlib, embedded Swift (no dylibs at all).

Reply:
You could also arrange that this can only be accessed by the SwiftLanguageRuntime, and then its loading will be gated on the Runtime loading, which already has to do this sort of computation.

Reply:
@jimingham I looked into that option, but AFAICT the SwiftLanguageRuntime is loaded regardless.
I.e. the first call to "load runtimes" will always load it, even if the relevant libraries haven't been loaded yet

Reply:
I found a private method in there that maybe we could expose...